<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
 <HEAD>
   <TITLE> [rabbitmq-discuss] Pika: Error when doing a stop_consuming() from a timeout callback
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:rabbitmq-discuss%40lists.rabbitmq.com?Subject=Re%3A%20%5Brabbitmq-discuss%5D%20Pika%3A%20Error%20when%20doing%20a%20stop_consuming%28%29%0A%20from%20a%20timeout%20callback&In-Reply-To=%3CBAA61C5E3C344A9BBA1E1542CA95FECD%40myyearbook.com%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="014230.html">
   <LINK REL="Next"  HREF="014232.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[rabbitmq-discuss] Pika: Error when doing a stop_consuming() from a timeout callback</H1>
    <B>Gavin M. Roy</B> 
    <A HREF="mailto:rabbitmq-discuss%40lists.rabbitmq.com?Subject=Re%3A%20%5Brabbitmq-discuss%5D%20Pika%3A%20Error%20when%20doing%20a%20stop_consuming%28%29%0A%20from%20a%20timeout%20callback&In-Reply-To=%3CBAA61C5E3C344A9BBA1E1542CA95FECD%40myyearbook.com%3E"
       TITLE="[rabbitmq-discuss] Pika: Error when doing a stop_consuming() from a timeout callback">gmr at myyearbook.com
       </A><BR>
    <I>Fri Jul 29 00:05:44 BST 2011</I>
    <P><UL>
        <LI>Previous message: <A HREF="014230.html">[rabbitmq-discuss] Pika: Error when doing a stop_consuming() from a timeout callback
</A></li>
        <LI>Next message: <A HREF="014232.html">[rabbitmq-discuss] Pika: Error when doing a stop_consuming() from a timeout callback
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#14231">[ date ]</a>
              <a href="thread.html#14231">[ thread ]</a>
              <a href="subject.html#14231">[ subject ]</a>
              <a href="author.html#14231">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>I'll take a look shortly, mind opening a issue on github for me to track with? 


On Thursday, July 28, 2011 at 7:04 PM, Christian Haintz wrote:

&gt;<i> On 29/7/11 24:54, Gavin M. Roy wrote:
</I>&gt;<i> &gt; 
</I>&gt;<i> &gt; On Thursday, July 28, 2011 at 6:45 PM, Christian Haintz wrote:
</I>&gt;<i> &gt; 
</I>&gt;<i> &gt; &gt; Hi,
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; I want to stop the consuming loop of a BlockingConnection after some
</I>&gt;<i> &gt; &gt; timeout. I used the following code. Producer is not running because I
</I>&gt;<i> &gt; &gt; want to simulate the timeout. Current pika version is used from github.
</I>&gt;<i> &gt; I'd not consider the master branch on Pika stable, which is one of the
</I>&gt;<i> &gt; reasons I've not tagged 0.9.6. I'd do a pip install of 0.9.5 and see if
</I>&gt;<i> &gt; that solves your problem.
</I>&gt;<i> &gt; 
</I>&gt;<i> &gt; Gavin
</I>&gt;<i> 
</I>&gt;<i> I tried it first with 0.9.5 but that leads to a different error which 
</I>&gt;<i> seams to be fixed in the master branch (used the same test code as in 
</I>&gt;<i> the previous e-mail):
</I>&gt;<i> 
</I>&gt;<i> --------8&lt;--------- snip ----------8&lt;-----------
</I>&gt;<i> python pika_test.py
</I>&gt;<i> Traceback (most recent call last):
</I>&gt;<i>  File &quot;pika_test.py&quot;, line 21, in &lt;module&gt;
</I>&gt;<i>  channel.start_consuming()
</I>&gt;<i>  File 
</I>&gt;<i> &quot;build/bdist.macosx-10.6-universal/egg/pika/adapters/blocking_connection.py&quot;, 
</I>&gt;<i> line 293, in start_consuming
</I>&gt;<i>  File 
</I>&gt;<i> &quot;build/bdist.macosx-10.6-universal/egg/pika/adapters/blocking_connection.py&quot;, 
</I>&gt;<i> line 103, in process_data_events
</I>&gt;<i>  File 
</I>&gt;<i> &quot;build/bdist.macosx-10.6-universal/egg/pika/adapters/blocking_connection.py&quot;, 
</I>&gt;<i> line 157, in process_timeouts
</I>&gt;<i> NameError: global name 'log' is not defined
</I>&gt;<i> --------8&lt;------- snip end --------8&lt;-----------
</I>&gt;<i> 
</I>&gt;<i> Best Regards,
</I>&gt;<i> Christian
</I>&gt;<i> 
</I>&gt;<i> 
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; --------8&lt;--------- snip ----------8&lt;-----------
</I>&gt;<i> &gt; &gt; import pika
</I>&gt;<i> &gt; &gt; connection = pika.BlockingConnection(pika.ConnectionParameters(
</I>&gt;<i> &gt; &gt; host='localhost'))
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; channel = connection.channel()
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; result = channel.queue_declare(exclusive=True)
</I>&gt;<i> &gt; &gt; callback_queue = result.method.queue
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; def on_response(self, ch, method, props, body):
</I>&gt;<i> &gt; &gt; #do some processing
</I>&gt;<i> &gt; &gt; channel.stop_consuming()
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; # normaly i do some publish here but that's not relevant
</I>&gt;<i> &gt; &gt; # to reproduce the error
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; channel.basic_consume(on_response, no_ack=True,
</I>&gt;<i> &gt; &gt; queue=callback_queue)
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; def panic():
</I>&gt;<i> &gt; &gt; channel.stop_consuming()
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; connection.add_timeout(1, panic)
</I>&gt;<i> &gt; &gt; channel.start_consuming()
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; --------8&lt;------- snip end --------8&lt;-----------
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; But I get the following error:
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; --------8&lt;--------- snip ----------8&lt;-----------
</I>&gt;<i> &gt; &gt; $ python pika_test.py
</I>&gt;<i> &gt; &gt; /Users/christian/devel/pika/pika/callback.py:69: UserWarning:
</I>&gt;<i> &gt; &gt; CallbackManager.add: Duplicate callback found for &quot;1:Basic.CancelOk&quot;
</I>&gt;<i> &gt; &gt; (self.__class__.__name__, prefix, key))
</I>&gt;<i> &gt; &gt; Traceback (most recent call last):
</I>&gt;<i> &gt; &gt; File &quot;pika_test.py&quot;, line 21, in &lt;module&gt;
</I>&gt;<i> &gt; &gt; channel.start_consuming()
</I>&gt;<i> &gt; &gt; File
</I>&gt;<i> &gt; &gt; &quot;/Users/christian/devel/pika/pika/adapters/blocking_connection.py&quot;, line
</I>&gt;<i> &gt; &gt; 307, in start_consuming
</I>&gt;<i> &gt; &gt; self.transport.connection.process_data_events()
</I>&gt;<i> &gt; &gt; File
</I>&gt;<i> &gt; &gt; &quot;/Users/christian/devel/pika/pika/adapters/blocking_connection.py&quot;, line
</I>&gt;<i> &gt; &gt; 104, in process_data_events
</I>&gt;<i> &gt; &gt; self.process_timeouts()
</I>&gt;<i> &gt; &gt; File
</I>&gt;<i> &gt; &gt; &quot;/Users/christian/devel/pika/pika/adapters/blocking_connection.py&quot;, line
</I>&gt;<i> &gt; &gt; 164, in process_timeouts
</I>&gt;<i> &gt; &gt; self._timeouts[timeout_id]['handler']()
</I>&gt;<i> &gt; &gt; File &quot;pika_test.py&quot;, line 18, in panic
</I>&gt;<i> &gt; &gt; channel.stop_consuming()
</I>&gt;<i> &gt; &gt; File
</I>&gt;<i> &gt; &gt; &quot;/Users/christian/devel/pika/pika/adapters/blocking_connection.py&quot;, line
</I>&gt;<i> &gt; &gt; 318, in stop_consuming
</I>&gt;<i> &gt; &gt; self.basic_cancel(consumer_tag)
</I>&gt;<i> &gt; &gt; File &quot;/Users/christian/devel/pika/pika/channel.py&quot;, line 333, in
</I>&gt;<i> &gt; &gt; basic_cancel
</I>&gt;<i> &gt; &gt; self._on_cancel_ok, [spec.Basic.CancelOk])
</I>&gt;<i> &gt; &gt; File
</I>&gt;<i> &gt; &gt; &quot;/Users/christian/devel/pika/pika/adapters/blocking_connection.py&quot;, line
</I>&gt;<i> &gt; &gt; 220, in rpc
</I>&gt;<i> &gt; &gt; self.send_method(method, None, wait)
</I>&gt;<i> &gt; &gt; File
</I>&gt;<i> &gt; &gt; &quot;/Users/christian/devel/pika/pika/adapters/blocking_connection.py&quot;, line
</I>&gt;<i> &gt; &gt; 249, in send_method
</I>&gt;<i> &gt; &gt; self.connection.process_data_events()
</I>&gt;<i> &gt; &gt; File
</I>&gt;<i> &gt; &gt; &quot;/Users/christian/devel/pika/pika/adapters/blocking_connection.py&quot;, line
</I>&gt;<i> &gt; &gt; 104, in process_data_events
</I>&gt;<i> &gt; &gt; self.process_timeouts()
</I>&gt;<i> &gt; &gt; File
</I>&gt;<i> &gt; &gt; &quot;/Users/christian/devel/pika/pika/adapters/blocking_connection.py&quot;, line
</I>&gt;<i> &gt; &gt; 164, in process_timeouts
</I>&gt;<i> &gt; &gt; self._timeouts[timeout_id]['handler']()
</I>&gt;<i> &gt; &gt; File &quot;pika_test.py&quot;, line 18, in panic
</I>&gt;<i> &gt; &gt; channel.stop_consuming()
</I>&gt;<i> &gt; &gt; File
</I>&gt;<i> &gt; &gt; &quot;/Users/christian/devel/pika/pika/adapters/blocking_connection.py&quot;, line
</I>&gt;<i> &gt; &gt; 318, in stop_consuming
</I>&gt;<i> &gt; &gt; self.basic_cancel(consumer_tag)
</I>&gt;<i> &gt; &gt; File &quot;/Users/christian/devel/pika/pika/channel.py&quot;, line 333, in
</I>&gt;<i> &gt; &gt; basic_cancel
</I>&gt;<i> &gt; &gt; self._on_cancel_ok, [spec.Basic.CancelOk])
</I>&gt;<i> &gt; &gt; File
</I>&gt;<i> &gt; &gt; &quot;/Users/christian/devel/pika/pika/adapters/blocking_connection.py&quot;, line
</I>&gt;<i> &gt; &gt; 209, in rpc
</I>&gt;<i> &gt; &gt; self._on_rpc_complete)
</I>&gt;<i> &gt; &gt; TypeError: 'NoneType' object is not iterable
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; What I am doing wrong? Are the timeout callbacks not capable of stopping
</I>&gt;<i> &gt; &gt; the consumer?
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; 
</I>&gt;<i> &gt; &gt; Regards,
</I>&gt;<i> &gt; &gt; Christian
</I>&gt;<i> &gt; &gt; _______________________________________________
</I>&gt;<i> &gt; &gt; rabbitmq-discuss mailing list
</I>&gt;<i> &gt; &gt; <A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">rabbitmq-discuss at lists.rabbitmq.com</A> (mailto:<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">rabbitmq-discuss at lists.rabbitmq.com</A>)
</I>&gt;<i> &gt; &gt; &lt;mailto:<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">rabbitmq-discuss at lists.rabbitmq.com</A>&gt;
</I>&gt;<i> &gt; &gt; <A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss</A>
</I>
-------------- next part --------------
An HTML attachment was scrubbed...
URL: &lt;<A HREF="http://lists.rabbitmq.com/pipermail/rabbitmq-discuss/attachments/20110728/4cc90584/attachment.htm">http://lists.rabbitmq.com/pipermail/rabbitmq-discuss/attachments/20110728/4cc90584/attachment.htm</A>&gt;
</PRE>










<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="014230.html">[rabbitmq-discuss] Pika: Error when doing a stop_consuming() from a timeout callback
</A></li>
	<LI>Next message: <A HREF="014232.html">[rabbitmq-discuss] Pika: Error when doing a stop_consuming() from a timeout callback
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#14231">[ date ]</a>
              <a href="thread.html#14231">[ thread ]</a>
              <a href="subject.html#14231">[ subject ]</a>
              <a href="author.html#14231">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">More information about the rabbitmq-discuss
mailing list</a><br>
</body></html>
